presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/TestIterativeOptimizer.java

```diff
@@ -0,0 +1,89 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.iterative;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.sql.planner.PlanNodeIdAllocator;
+import com.facebook.presto.sql.planner.StatsRecorder;
+import com.facebook.presto.sql.planner.SymbolAllocator;
+import com.facebook.presto.sql.planner.optimizations.PlanOptimizer;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.testing.LocalQueryRunner;
+import com.facebook.presto.tpch.TpchConnectorFactory;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+import java.util.Optional;
+
+import static com.facebook.presto.spi.StandardErrorCode.OPTIMIZER_TIMEOUT;
+import static com.facebook.presto.testing.TestingSession.testSessionBuilder;
+import static org.testng.Assert.assertEquals;
+
+public class TestIterativeOptimizer
+{
+    LocalQueryRunner queryRunner;
+
+    @BeforeClass
+    public void setUp()
+    {
+        Session.SessionBuilder sessionBuilder = testSessionBuilder()
+                .setCatalog("local")
+                .setSchema("tiny")
+                .setSystemProperty("task_concurrency", "1")
+                .setSystemProperty("iterative_optimizer_enabled", "true");
+
+        this.queryRunner = new LocalQueryRunner(sessionBuilder.build());
+
+        queryRunner.createCatalog(queryRunner.getDefaultSession().getCatalog().get(),
+                new TpchConnectorFactory(1),
+                ImmutableMap.of());
+    }
+
+    @AfterClass
+    public void tearDown()
+    {
+        queryRunner.close();
+    }
+
+    @Test(timeOut = 1000)
+    public void optimizerTimeoutsOnNonConvergingPlan()
+    {
+        PlanOptimizer optimizer = new IterativeOptimizer(new StatsRecorder(), ImmutableSet.of(new NonConvergingRule()), 1);
+
+        try {
+            queryRunner.inTransaction(transactionSession -> {
+                queryRunner.createPlan(transactionSession, "SELECT * FROM nation", ImmutableList.of(optimizer));
+                return null;
+            });
+        }
+        catch (PrestoException ex) {
+            assertEquals(ex.getErrorCode(), OPTIMIZER_TIMEOUT.toErrorCode());
+        }
+    }
+
+    private static class NonConvergingRule
+            implements Rule
+    {
+        @Override
+        public Optional<PlanNode> apply(PlanNode node, Lookup lookup, PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator)
+        {
+            return Optional.of(node);
```

Comment:
This may not be a good way to produce a non-converging rule in the long term. I could see us updating the optimizer to ignore the result of the transformation if it's exactly the same object that was passed in (I though I had added this already, but maybe it was in some prototype code I was working on).